### PR TITLE
Premint - get array of supported premint versions

### PR DESCRIPTION
**Pull request by oveddan**

### CODE REVIEW

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #278** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/278" target="_blank">Graphite</a> 
    * **PR #297** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/297" target="_blank">Graphite</a> 
    * **PR #290** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/290" target="_blank">Graphite</a> 
      * **PR #291** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/291" target="_blank">Graphite</a> 
        * **PR #262** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/262" target="_blank">Graphite</a> 
          * **PR #294** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/294" target="_blank">Graphite</a> 
            * **PR #295** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/295" target="_blank">Graphite</a> 
              * **PR #289** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/289" target="_blank">Graphite</a> 
                * **PR #299** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/299" target="_blank">Graphite</a> 
                  * **PR #303** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/303" target="_blank">Graphite</a> 
                    * **PR #309** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/309" target="_blank">Graphite</a>  👈
          * **PR #273** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/273" target="_blank">Graphite</a> 
            * **PR #288** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/288" target="_blank">Graphite</a> 
              * **PR #284** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/284" target="_blank">Graphite</a> 
            * **PR #280** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/280" target="_blank">Graphite</a> 
    * **PR #283** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/283" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-protocol/309?utm_source=stack-comment).

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #278** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/278" target="_blank">Graphite</a> 
    * **PR #297** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/297" target="_blank">Graphite</a> 
    * **PR #290** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/290" target="_blank">Graphite</a> 
      * **PR #291** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/291" target="_blank">Graphite</a> 
        * **PR #262** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/262" target="_blank">Graphite</a> 
          * **PR #294** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/294" target="_blank">Graphite</a> 
            * **PR #295** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/295" target="_blank">Graphite</a> 
              * **PR #289** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/289" target="_blank">Graphite</a> 
                * **PR #299** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/299" target="_blank">Graphite</a> 
                  * **PR #303** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/303" target="_blank">Graphite</a> 
                    * **PR #309** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/309" target="_blank">Graphite</a>  👈
          * **PR #273** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/273" target="_blank">Graphite</a> 
            * **PR #288** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/288" target="_blank">Graphite</a> 
              * **PR #284** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/284" target="_blank">Graphite</a> 
            * **PR #280** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/280" target="_blank">Graphite</a> 
    * **PR #283** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/283" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-protocol/309?utm_source=stack-comment).

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #278** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/278" target="_blank">Graphite</a> 
    * **PR #297** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/297" target="_blank">Graphite</a> 
    * **PR #290** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/290" target="_blank">Graphite</a> 
      * **PR #291** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/291" target="_blank">Graphite</a> 
        * **PR #262** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/262" target="_blank">Graphite</a> 
          * **PR #294** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/294" target="_blank">Graphite</a> 
            * **PR #295** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/295" target="_blank">Graphite</a> 
              * **PR #289** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/289" target="_blank">Graphite</a> 
                * **PR #299** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/299" target="_blank">Graphite</a> 
                  * **PR #303** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/303" target="_blank">Graphite</a> 
                    * **PR #309** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/309" target="_blank">Graphite</a>  👈
          * **PR #273** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/273" target="_blank">Graphite</a> 
            * **PR #288** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/288" target="_blank">Graphite</a> 
              * **PR #284** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/284" target="_blank">Graphite</a> 
            * **PR #280** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/280" target="_blank">Graphite</a> 
    * **PR #283** <a href="https://app.graphite.dev/github/pr/ourzora/zora-protocol/283" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-protocol/309?utm_source=stack-comment).

File: packages/1155-contracts/src/delegation/ZoraCreator1155Attribution.sol [hunk 1]
```diff
@@ -95,6 +95,12 @@ library ZoraCreator1155Attribution {
     string internal constant VERSION_2 = "2";
     bytes32 internal constant HASHED_VERSION_2 = keccak256(bytes(VERSION_2));
 
+    function allVersions() internal pure returns (string[] memory versions) {
+        versions = new string[](2);
```
Comment (kulkarohan) [inline packages/1155-contracts/src/delegation/ZoraCreator1155Attribution.sol:99]: does this need to be in a separate function from `supportedPremintSignatureVersions()`?

File: packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol [hunk 2]
```diff
@@ -214,21 +214,28 @@ contract ZoraCreator1155PremintExecutorImpl is
...
-    /// @return The version of the premint signature that the contract supports.  If it doesn't support premint
-    /// returns 0
-    function supportedPremintSignatureVersion(address contractAddress) external view returns (string memory) {
+    /// @return versions The versions of the premint signature that the contract supports.  If contract hasn't been created yet,
+    /// assumes that when it will be created it will support the latest versions of the signatures, so the function returns all versions.
+    function supportedPremintSignatureVersions(address contractAddress) external view returns (string[] memory versions) {
+        // if contract hasn't been created yet, assume it will be created with the latest version
+        // and thus supports all versions of the signature
+        if (contractAddress.code.length == 0) {
+            return ZoraCreator1155Attribution.allVersions();
```
Comment (kulkarohan) [inline packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol:225]: should we return all versions here or just an array of 1 with the latest?

Comment (oveddan) [reply packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol:225]: good q. if the contract hasn't been created yet, and a premint v1 signature has been created for that address, then the new contract can be created with the v1 version, so in theory both v1 and v2 would work

Comment (oveddan) [reply packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol:225]: i do see how this is wierd that we are returning all versions if the contract hasn't been created yet.  may be just simpler to revert.  what do you think @iainnash @kulkarohan ?

**Review by IsabellaSmallcombe — APPROVED** [review]

File: packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol (hunk 2 above)
Comment (iainnash) [reply packages/1155-contracts/src/delegation/ZoraCreator1155PremintExecutorImpl.sol:225]: I don't mind all versions tbh, feels cleaner.

File: packages/1155-contracts/src/nft/ZoraCreator1155Impl.sol [hunk 3]
```diff
@@ -752,8 +752,8 @@ contract ZoraCreator1155Impl is
         return _getImplementation();
     }
 
-    function supportedPremintSignatureVersion() external pure returns (string memory) {
-        return ZoraCreator1155Attribution.VERSION_2;
+    function supportedPremintSignatureVersions() external pure returns (string[] memory) {
+        return DelegatedTokenCreation.supportedPremintSignatureVersions();
```
Comment (iainnash) [inline packages/1155-contracts/src/nft/ZoraCreator1155Impl.sol:756]: How much codespace does this approach use? What about using a bitmap :P?

**Review by iainnash — APPROVED** [review]
I wonder if we can come up with a more efficient way to do this codespace on the main nft contract. Bitmap feels too obtuse but interested in ideas.

File: packages/1155-contracts/src/delegation/ZoraCreator1155Attribution.sol (hunk 1 above)
Comment (oveddan) [reply packages/1155-contracts/src/delegation/ZoraCreator1155Attribution.sol]: good q - its called via an external lib call, which saves codespace.

Comment (oveddan) [reply packages/1155-contracts/src/delegation/ZoraCreator1155Attribution.sol]: ive actually got rid of this method and extra level of indirection and just flattened it into:

`DelegatedTokenCreation.supportedPremintSignatureVersions`

**Pull request merged**